Resolve #1222 -- Implement OnDash and OnFinishDash events

Dawiss2 opened the pull request:
- Implemented OnDash event
- Implemented OnFinishDash event
- Player is now unable to change dash target while dashing

Refs #1222

Dawiss2 committed:
Resolve #1222 -- Implement OnDash and OnFinishDash events

- Implemented OnDash event
- Implemented OnFinishDash event
- Player is now unable to change dash target while dashing

Refs #1222

lzardy commented on GameServerLib/API/ApiEventManager.cs:
```diff
@@ -37,6 +37,8 @@
 [OnLevelUp]
 [OnLevelUpSpell]
 [OnMiss]
+[OnFinishDash]
```
Remove these and use `OnMoveEnd`, `OnMoveSuccess` and `OnMoveFailure`.
An `OnDash` is unnecessary as all units have MovementParameters which can be checked.

Dawiss2 replied:
Soo i should remove OnDash and OnFinishDash and create OnMoveEnd, OnMoveSuccess and OnMoveFailure events instead? And publish them from Move functions instead of Dash? Or keep OnFinishDash and just do OnMove events instead OnDash? Sorry but i don't understand.

lzardy commented on GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs:
```diff
@@ -830,7 +829,7 @@ public void SetSpellToCast(ISpell s, Vector2 location, IAttackableUnit unit = nu
         public void SetTargetUnit(IAttackableUnit target, bool networked = false)
         {
-            if (target == null || target.IsDead || !target.Status.HasFlag(StatusFlags.Targetable))
+            if (target == null || target.IsDead || !target.Status.HasFlag(StatusFlags.Targetable) || MovementParameters != null)
```
Remove this, add a `Target GameObject` to the MovementParameters and use that for the targeted forced movements (dashes).

lzardy commented on GameServerLib/GameObjects/AttackableUnits/AttackableUnit.cs:
```diff
@@ -1487,14 +1487,22 @@ public void DashToLocation(Vector2 endPos, float dashSpeed, string animation = "
+            }
+            if(state)
+            {
+                ApiEventManager.OnDash.Publish(this);
```
Remove this.

lzardy reviewed (changes requested):
Quality of changes should be improved.

danil179 commented on GameServerLib/GameObjects/AttackableUnits/AttackableUnit.cs:
```diff
@@ -1487,14 +1487,22 @@ public void DashToLocation(Vector2 endPos, float dashSpeed, string animation = "
         /// <param name="state">State to set. True = dashing, false = not dashing.</param>
         /// TODO: Implement ForcedMovement methods and enumerators to handle different kinds of dashes.
-        public virtual void SetDashingState(bool state)
+        public virtual void SetDashingState(bool state, bool wasCancelled = false)
```
I think it is better to change the state to enum. Avoid adding additional booleans as it will complicate many things in the future.

Dawiss2 replied:
Done i think, not sure if i did it in a right way.

danil179 reviewed (changes requested):
I suggest to change the dashing state to include more than 2 states (true/false).

Dawiss2 committed:
Resolve #1224 -- Changed the SetDashingState to use enums.

- Changed SetDashingState to use enum instead of bool
- Added ForceMovementState enum

refs #1224

lzardy commented:
This PR has been superseded by #1301
